File: codes/cable/Twisted-wire pairs/TWP_far_field/TWP_far_field.m
clear;clc;
%% Free Space charachteristic
u0 = 4*pi*1e-7; % permeability. 
epsi = 8.85e-12; % permittivity.
c0 = 299792458; % speed of light in free space.

%% Twisted bundel parameters
s = 0.25e-2;
h = 1e-2;
rw = 1.46e-9;
p = 5e-2;
Lz = 1;
alpha = ((s/2)^2+(p/(2*pi))^2)^(-1/2);
L = 2*pi*Lz/alpha/p;

%% Averaged_per_unit_length_parameters
Ra = 50; Rb = 50; Rc = 100;
l_avg  = u0/(2*pi)*log(2*h/rw);
lm_avg = u0/(2*pi)*(log(2*h/s)+s^2/(16*h^2));
L_avg  = [l_avg,lm_avg;lm_avg,l_avg];
C_avg  = (1/(c0^2))*L_avg;
Zc = c0*L_avg;

Zl = [Ra+Rc,Rc;Rc,Rb+Rc];
Zs = [Ra+Rc,Rc;Rc,Rb+Rc];

Zc_vr = c0*u0/(2*pi)*(log(2*h/rw)-1);  
L_vr = h*Zc_vr/c0;
Zc_dr = c0*u0/(2*pi)*log(s/rw);

%% Import Data
I0_Feko = load('endfire.dat');

%% Plane-wave Characterization
E0 = 1;
%Side-fire
% thetaE = deg2rad(0);
% thetaP = deg2rad(0);   
% phiP = deg2rad(0);
%End-Fire
thetaE = deg2rad(90);
thetaP = deg2rad(90);   
phiP = deg2rad(-90);
%Broad-Side
% thetaE = deg2rad(90);
% thetaP = deg2rad(90);   
% phiP = deg2rad(0);
ex = sin(thetaE)*sin(thetaP); 
ey = -sin(thetaE)*cos(thetaP)*cos(phiP)-cos(thetaE)*sin(phiP); 
ez = -sin(thetaE)*cos(thetaP)*sin(phiP)+cos(thetaE)*cos(phiP);

%% Twisting geometrical parameters
acc = 1e-5;
l = 0:acc:L;
for i = 1:2
    x  = h+(cos(alpha*l).*(s/2)*((-1)^(i-1)));    
    dx = -(alpha*sin(alpha*l).*(s/2)*((-1)^(i-1)));
    y  = sin(alpha*l).*(s/2)*((-1).^(i-1));      
    dy = alpha*cos(alpha*l).*(s/2)*(-1)^(i-1);
    z  = alpha*l.*p/(2*pi);                                             
    dz = alpha*p/(2*pi);

    %% Frequency loop 
    index = 0; 
    for f = linspace(1e6,1e9,1000)
        index = index+1;
        k = 2*pi*f/c0;  % The wave number.
        kx=-k*cos(thetaP);
        ky=-k*sin(thetaP)*cos(phiP);
        kz=-k*sin(thetaP)*sin(phiP);
        Ex=2*E0*ex*cos(kx*x).*exp(-1j*(ky*y+kz*z));     
        Ey=-2j*E0*ey*sin(kx*x).*exp(-1j*(ky*y+kz*z)); 
        Ez=-2j*E0*ez*sin(kx*x).*exp(-1j*(ky*y+kz*z)); 
        
        %% VSLl and VSRl
        VF = Ex.*dx+Ey.*dy+Ez.*dz;
        SNC1 = sin(k*(l-L))/sin(k*L);
        SNC2 = sin(k*l)/sin(k*L);
        vSLl = trapz(l,SNC1.*VF);   
        VSLl(index) = vSLl;
        vSRl = trapz(l,SNC2.*VF);   
        VSRl(index) = vSRl;
        
        %% VSLv and VSRv
        vSLv = -2*E0*ex*sin(kx*x(1))/kx;
        VSLv(index) = vSLv;
        vSRv = -2*E0*ex*sin(kx*x(length(x))).*exp(-1j*(ky*y(length(y))+kz*Lz))/kx;
        VSRv(index) = vSRv;

        %% Riser Current   
        VarName_L1 = 0; VarName_R1 = 0;
        Zrr = Zc_vr*(0+1j*Zc_vr*tan(k*h))/(Zc_vr+1j*0*tan(k*h))+Rc;
        Ivr0 = -1j*k*h^2*E0*ex*exp(-1j*kz*0)/Zc_vr-1j*k*s^2*(VarName_L1)/(8*Zc_dr);  
        IvrL = -1j*k*h^2*E0*ex*exp(-1j*kz*Lz)/Zc_vr-1j*k*s^2*(VarName_R1)/(8*Zc_dr); 
        Idr0 = -1j*k*(s/2)^2*E0*ex*exp(-1j*kz*0)/Zc_dr; 
        IdrL = -1j*k*(s/2)^2*E0*ex*exp(-1j*kz*Lz)/Zc_dr;
        vris0 = Zrr*Ivr0+(Ra*((-1)^(i+1)+1)/2-Rb*((-1)^(i)+1)/2)*Idr0;   
        Vris0(index) = vris0;  
        vrisL = Zrr*IvrL+(Ra*((-1)^(i+1)+1)/2-Rb*((-1)^(i)+1)/2)*IdrL;   
        VrisL(index) = vrisL; 
    end
    VVSL = VSLl+VSLv;    
    VVSR = VSRl+VSRv; 
    VSL(i,:) = VVSL;    
    VSR(i,:) = VVSR; 
end

index = 0;
for f = linspace(1e6,1e9,1000)
    index = index+1; 
    k = 2*pi*f/c0;
    %% Frequqncy Dependent Matrices
    PHI11 = cos(k*L)*eye(2);    
    PHI12 = -1j*Zc*sin(k*L);
    PHI21 = -1j*(Zc^-1)*sin(k*L);   
    PHI22 = cos(k*L)*eye(2);
    Zr = Zc_vr*(0+1j*Zc_vr*tan(k*h))/(Zc_vr+1j*0*tan(k*h));
    Zl1 = Zl+Zr*ones(2);
    Zs1 = Zl1;
    D = PHI11*Zs+Zl*PHI22-Zl*PHI21*Zs-PHI12;
    I0(:,index) = (D^-1)*((Zs*PHI21-PHI11)*VSL(:,index)+VSR(:,index));
end
figure();
f = linspace(1e6,1e9,1000);
semilogx(f,20*log10(abs(I0(1,:)+I0(2,:))),f,20*log10(abs((I0(1,:)-I0(2,:)))/2),'linewidth',2);
hold on
semilogx(I0_Feko(:,1),I0_Feko(:,2),'--','linewidth',2);
hold on
semilogx(I0_Feko(:,1),I0_Feko(:,3),'--','linewidth',2);
title('End-Fire');
legend('CM, analytical','DM, analytical','CM, Feko (MoM)','DM, Feko (MoM)');
xlabel('Frequency (Hz)');
ylabel('Modal Current (dBA)');

File: codes/TWP_far_field.py
import numpy as np
import matplotlib.pyplot as plt

def TWP_far_field(s, h, rw, p, Lz, fmin, fmax, Np):
    """计算双绞线远场特性
    
    Args:
        s (float): 导线间距
        h (float): 高度
        rw (float): 导线半径
        p (float): 节距
        Lz (float): 总长度
        fmin (float): 最小频率
        fmax (float): 最大频率
        Np (int): 频率点数
    
    Returns:
        tuple: (CMC, DMC, f) - 共模电流、差模电流和频率数组
    """
    # 自由空间特性
    u0 = 4 * np.pi * 1e-7    # 磁导率
    epsi = 8.85e-12          # 介电常数
    c0 = 299792458           # 真空光速

    # 扭绞束参数
    alpha = ((s/2)**2 + (p/(2*np.pi))**2)**(-1/2)
    L = 2*np.pi*Lz/alpha/p

    # 平均每单位长度参数
    Ra, Rb, Rc = 50, 50, 100
    l_avg = u0/(2*np.pi) * np.log(2*h/rw)
    lm_avg = u0/(2*np.pi) * (np.log(2*h/s) + s**2/(16*h**2))
    L_avg = np.array([[l_avg, lm_avg], [lm_avg, l_avg]])
    C_avg = (1/(c0**2)) * L_avg
    Zc = c0 * L_avg

    Zl = np.array([[Ra+Rc, Rc], [Rc, Rb+Rc]])
    Zs = Zl.copy()

    Zc_vr = c0*u0/(2*np.pi) * (np.log(2*h/rw) - 1)
    L_vr = h*Zc_vr/c0
    Zc_dr = c0*u0/(2*np.pi) * np.log(s/rw)

    # 平面波特性
    E0 = 1
    # End-Fire configuration
    thetaE = np.deg2rad(90)
    thetaP = np.deg2rad(90)
    phiP = np.deg2rad(-90)

    ex = np.sin(thetaE)*np.sin(thetaP)
    ey = -np.sin(thetaE)*np.cos(thetaP)*np.cos(phiP)-np.cos(thetaE)*np.sin(phiP)
    ez = -np.sin(thetaE)*np.cos(thetaP)*np.sin(phiP)+np.cos(thetaE)*np.cos(phiP)

    # 扭绞几何参数
    acc = 1e-5
    l = np.arange(0, L+acc, acc)
    frequencies = np.linspace(fmin, fmax, Np)
    VSL = np.zeros((2, Np), dtype=complex)
    VSR = np.zeros((2, Np), dtype=complex)

    for i in range(2):
        x = h + (np.cos(alpha*l)*(s/2)*((-1)**(i)))
        dx = -(alpha*np.sin(alpha*l)*(s/2)*((-1)**(i)))
        y = np.sin(alpha*l)*(s/2)*((-1)**(i))
        dy = alpha*np.cos(alpha*l)*(s/2)*(-1)**(i)
        z = alpha*l*p/(2*np.pi)
        dz = alpha*p/(2*np.pi)

        for idx, f in enumerate(frequencies):
            k = 2*np.pi*f/c0
            kx = -k*np.cos(thetaP)
            ky = -k*np.sin(thetaP)*np.cos(phiP)
            kz = -k*np.sin(thetaP)*np.sin(phiP)

            Ex = 2*E0*ex*np.cos(kx*x)*np.exp(-1j*(ky*y+kz*z))
            Ey = -2j*E0*ey*np.sin(kx*x)*np.exp(-1j*(ky*y+kz*z))
            Ez = -2j*E0*ez*np.sin(kx*x)*np.exp(-1j*(ky*y+kz*z))

            VF = Ex*dx + Ey*dy + Ez*dz
            SNC1 = np.sin(k*(l-L))/np.sin(k*L)
            SNC2 = np.sin(k*l)/np.sin(k*L)

            vSLl = np.trapezoid(SNC1*VF, l)
            vSRl = np.trapezoid(SNC2*VF, l)

            vSLv = -2*E0*ex*np.sin(kx*x[0])/kx
            vSRv = -2*E0*ex*np.sin(kx*x[-1])*np.exp(-1j*(ky*y[-1]+kz*Lz))/kx

            VSL[i,idx] = vSLl + vSLv
            VSR[i,idx] = vSRl + vSRv

    # 频率循环计算
    I0 = np.zeros((2, Np), dtype=complex)
    for idx, f in enumerate(frequencies):
        k = 2*np.pi*f/c0
        
        PHI11 = np.cos(k*L)*np.eye(2)
        PHI12 = -1j*Zc*np.sin(k*L)
        PHI21 = -1j*np.linalg.inv(Zc)*np.sin(k*L)
        PHI22 = np.cos(k*L)*np.eye(2)
        
        Zr = Zc_vr*(0+1j*Zc_vr*np.tan(k*h))/(Zc_vr+1j*0*np.tan(k*h))
        Zl1 = Zl + Zr*np.ones((2,2))
        Zs1 = Zl1.copy()
        
        D = PHI11@Zs + Zl@PHI22 - Zl@PHI21@Zs - PHI12
        I0[:,idx] = np.linalg.solve(D, (Zs@PHI21 - PHI11)@VSL[:,idx] + VSR[:,idx])

    # 计算远场
    CMC = I0[0,:] + I0[1,:]
    DMC = (I0[0,:] - I0[1,:])/2

    return CMC, DMC, frequencies

if __name__ == "__main__":
    # 测试参数
    s = 0.25e-2   # 导线间距
    h = 1e-2      # 高度
    rw = 1.46e-9  # 导线半径
    p = 5e-2      # 节距
    Lz = 1        # 总长度
    fmin = 1e6    # 最小频率
    fmax = 1e9    # 最大频率
    Np = 201      # 频率点数

    CMC, DMC, f = TWP_far_field(s, h, rw, p, Lz, fmin, fmax, Np)

    # 绘制结果
    plt.figure()
    plt.semilogx(f, 20*np.log10(np.abs(CMC)), 
                 f, 20*np.log10(np.abs(DMC)), 
                 linewidth=2)
    plt.title('End-Fire')
    plt.legend(['CM, analytical', 'DM, analytical'])
    plt.xlabel('Frequency (Hz)')
    plt.ylabel('Modal Current (dBA)')
    plt.grid(True)
    plt.show()